plugin.rb

```diff
@@ -25,101 +19,126 @@ class Engine < ::Rails::Engine
+      hash = $redis.hgetall(get_redis_key(type, id))
+      original_hash_size = hash.length
+
+      any_changes = false
+
+      # Delete entries older than 20 seconds
+      hash.each do |user_id, time|
+        if Time.zone.now - Time.parse(time) >= 20
+          any_changes ||= remove(type, id, user_id)
+        end
+      end
+
+      return any_changes
     end
 
-    MessageBus.publish("/presence-writing-#{msg.data['channel_id']}", { users: users, drafts: drafts })
   end
 
   require_dependency "application_controller"
 
   class Presence::PresencesController < ::ApplicationController
-    requires_plugin PLUGIN_NAME
+    before_filter :ensure_logged_in
 
-    def add
-      $redis.hset(redis_channel, current_user.id, Time.zone.now)
-      delete_old
-      MessageBus.publish('/writing-reply', {channel_id: params[:id], user_id: current_user.id.to_s})
-      render json: { subscribed: true }
-    end
+    def publish
+      data = params.permit(:response_needed,
+              current: [:compose_state, :action, :topic_id, :post_id],
+              previous: [:compose_state, :action, :topic_id, :post_id]
+              )
 
-    def remove
-      $redis.hdel(redis_channel, current_user.id)
-      MessageBus.publish('/writing-reply', {channel_id: params[:id], user_id: current_user.id.to_s})
-      render json: { subscribed: false }
-    end
+      if data[:previous] &&
+          data[:previous][:compose_state] == 'open' &&
+          data[:previous][:action].in?(['edit', 'reply'])
 
-    def alive
-      $redis.hset(redis_channel, current_user.id, Time.zone.now)
-      value = delete_old
-      hash = value[0]
-      original_hash_size = value[1]
+        type = data[:previous][:post_id] ? 'post' : 'topic'
+        id = data[:previous][:post_id] ? data[:previous][:post_id] : data[:previous][:topic_id]
 
-      # Publish the new hash if there were any changes
-      if original_hash_size != hash.length
-        MessageBus.publish('/writing-reply', {channel_id: params[:id], user_id: current_user.id.to_s})
+        any_changes = false;
+        any_changes ||= Presence::PresenceManager.remove(type, id, current_user.id)
+        any_changes ||= Presence::PresenceManager.cleanup(type, id)
+
+        users = Presence::PresenceManager.publish(type, id) if any_changes
       end
 
-      render json: {alive: true}
-    end
+      if data[:current] &&
+          data[:current][:compose_state] == 'open' &&
+          data[:current][:action].in?(['edit', 'reply'])
 
-    def delete_old
-      hash = $redis.hgetall(redis_channel)
-      original_hash_size = hash.length
+        type = data[:current][:post_id] ? 'post' : 'topic'
+        id = data[:current][:post_id] ? data[:current][:post_id] : data[:current][:topic_id]
 
-      # Delete entries older than 20 seconds
-      hash.each do |user, time|
-        if Time.zone.now - Time.parse(time) >= 20
-          $redis.hdel(redis_channel, user)
-          hash.delete(user)
+        any_changes = false;
```

Comment:
Oh look a semicolon 😉

Reply:
😂 oh dear. Surprised rubocop didn’t moan about that one